### PR TITLE
Implement Hop-Based TTL + Bugfixes

### LINKED ISSUE #12: TTL is Time based rather than Hop based
In PUBLISH messages the TTL is currently time-based. In the [Quasar paper](http://research.microsoft.com/en-us/um/people/saikat/pub/iptps08-quasar.pdf) that this library is based on the routing TTL is hop-based like the TTL in an IP header and gets decremented for each hop that the messages is routed.

Using a time based TTL means the life of a message and it's distance traveled is heavily dependent on the current latency of the network. It also means in a low latency network condition the routing to the random node could cause the message to be routed to every single node in the network which seems to bypass the purpose of the algorithm.

### LINKED ISSUE #15: Random relay not working
### Expected Behavior

I would expect that random relay will take 3 random nodes and send them the message.
(random relay is activated here: https://github.com/Storj/core/blob/master/lib/network/index.js#L448)

### Actual Behavior

My renter is uploading nonstop but my farmer don't received any contracts. The reason for this is random relay. I have 10 nodes on my contact list. I generate random key but end up with the same random neighbor all the time.

I don't see a random distribution and I would expect at least 2 PUBLISH messages and not only 1 message. With only 1 message every "message previously seen" is a dead end.

```
grep 'Random\|PUBLISH' .pm2/logs/storj1-out-0.log
{"level":"info","message":"PUBLISH topic: 0f01020202","timestamp":"2016-10-22T21:32:28.891Z"}
{"level":"debug","message":"Random Key: c7c8737aaca94244c119aaed443d06da462f2de1","timestamp":"2016-10-22T21:32:28.954Z"}
{"level":"debug","message":"Random Neighbor: 012c.storjlt.xyz:5001","timestamp":"2016-10-22T21:32:28.954Z"}
{"level":"info","message":"sending PUBLISH message to {\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"012c.storjlt.xyz\",\"port\":5001,\"nodeID\":\"4a19bdae50da73812ca48301318f147ca8496708\",\"lastSeen\":1477171948955}","timestamp":"2016-10-22T21:32:28.955Z"}
[...]
{"level":"info","message":"PUBLISH topic: 0f01020202","timestamp":"2016-10-22T21:32:30.863Z"}
{"level":"debug","message":"Random Key: 0e561ee32858d71e1d1cf65f64fbf2b181d5540f","timestamp":"2016-10-22T21:32:30.926Z"}
{"level":"debug","message":"Random Neighbor: 012c.storjlt.xyz:5001","timestamp":"2016-10-22T21:32:30.927Z"}
{"level":"info","message":"sending PUBLISH message to {\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"012c.storjlt.xyz\",\"port\":5001,\"nodeID\":\"4a19bdae50da73812ca48301318f147ca8496708\",\"lastSeen\":1477171950927}","timestamp":"2016-10-22T21:32:30.927Z"}
{"level":"info","message":"PUBLISH topic: 0f01020202","timestamp":"2016-10-22T21:32:31.038Z"}
{"level":"debug","message":"Random Key: 956162ed8faa675ea40d04d6d5f28c96cbdb8a06","timestamp":"2016-10-22T21:32:31.100Z"}
{"level":"debug","message":"Random Neighbor: 012c.storjlt.xyz:5001","timestamp":"2016-10-22T21:32:31.101Z"}
{"level":"info","message":"sending PUBLISH message to {\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"012c.storjlt.xyz\",\"port\":5001,\"nodeID\":\"4a19bdae50da73812ca48301318f147ca8496708\",\"lastSeen\":1477171951101}","timestamp":"2016-10-22T21:32:31.101Z"}
{"level":"info","message":"PUBLISH topic: 0f01020202","timestamp":"2016-10-22T21:32:32.967Z"}
{"level":"debug","message":"Random Key: 90eada84e6c310495f48fabd99d64d4506a89c91","timestamp":"2016-10-22T21:32:33.033Z"}
{"level":"debug","message":"Random Neighbor: 012c.storjlt.xyz:5001","timestamp":"2016-10-22T21:32:33.034Z"}
{"level":"info","message":"sending PUBLISH message to {\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"012c.storjlt.xyz\",\"port\":5001,\"nodeID\":\"4a19bdae50da73812ca48301318f147ca8496708\",\"lastSeen\":1477171953034}","timestamp":"2016-10-22T21:32:33.034Z"}
```

```
{"level":"info","message":"received FIND_NODE from {\"userAgent\":\"4.0.6\",\"protocol\":\"0.9.1\",\"address\":\"STORJ-CLUS-083.storj.space\",\"port\":4012,\"nodeID\":\"2a8a5eebc6826181f2b5bbb0340daa7f52f037cb\",\"lastSeen\":1477061301584}","timestamp":"2016-10-22T22:32:25.578Z"}
{"level":"debug","message":"sending 2a8a5eebc6826181f2b5bbb0340daa7f52f037cb nearest 10 contacts","timestamp":"2016-10-22T22:32:25.579Z"}
```

### LINKED ISSUE #16: Messages not relayed
### Expected Behavior

Relay PUBLISH messages to 3 neighbors that have not seen the message.

### Actual Behavior

Get 3 close neighbors and remove them if they have seen the message. Most of the time we will relay it only to one neighbor and sometime all 3 neighbors already received the message. This is a dead end for the message with only 4 hops.

I have at least 10 contacts on my list. 6 or 7 have not seen the message. There is no reason to drop the message. 
```
{"level":"info","message":"Neighbors unfiltered: [{\"userAgent\":\"4.0.6\",\"protocol\":\"0.9.1\",\"address\":\"104.196.221.141\",\"port\":8470,\"nodeID\":\"593844dc7f0076a1aeda9a6b9788af17e67c1052\",\"lastSeen\":1477226032550},{\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"STORJ-CLUS-113.storj.space\",\"port\":4074,\"nodeID\":\"6bd8a6fabbed1e7dace9fe20dc9f0d73f2a07274\",\"lastSeen\":1477226033577},{\"userAgent\":\"4.1.0\",\"protocol\":\"0.9.1\",\"address\":\"STORJ-CLUS-085.storj.space\",\"port\":4022,\"nodeID\":\"1392dbfa2641a6cced0dfadacbccd5949ccbc4a8\",\"lastSeen\":1477226033127}]","timestamp":"2016-10-23T12:33:53.647Z"}
{"level":"info","message":"Publisher list: [\"a69d4f2e3e97e4f26ece3d437b2825c5d7fed9e6\",\"1392dbfa2641a6cced0dfadacbccd5949ccbc4a8\",\"593844dc7f0076a1aeda9a6b9788af17e67c1052\",\"6bd8a6fabbed1e7dace9fe20dc9f0d73f2a07274\"]","timestamp":"2016-10-23T12:33:53.647Z"}
{"level":"info","message":"Neighbors filtered: []","timestamp":"2016-10-23T12:33:53.647Z"}
```

### PR DESCRIPTION
**Pull request by bookchin**
* Close #12 
* Close #15 
* Close #16

### CODE REVIEW

**Comment by yuilleb** [comment]
hey @bookchin I'm not seeing anywhere that the ttl parameter is validated as a positive integer? Is this being done? Hopefully I'm just being dense.

**Comment by bookchin** [comment]
Yep, didn't add that, but definitely should.

**Comment by yuilleb** [comment]
Anyway to remove the 1.0.0 release while you're working on this? we don't want people installing this as it has a dos vector in it.

**Comment by bookchin** [comment]
Storj-lib will not automatically upgrade, since this is backwards incompatible protocol change. I'll fix it shortly and issue an update

**Comment by yuilleb** [comment]
OK great. Anyway good work getting this code done!

File: lib/quasar.js [hunk 1]
```diff
@@ -193,16 +193,17 @@ Quasar.prototype._applyAttenuatedBloomFilterUpdate = function(atbf) {
  * @param {Function} callback
  */
 Quasar.prototype._sendPublish = function(contact, params, callback) {
-  var tnow = Date.now();
-
   // check to make sure the message we are sending hasn't expired:
-  if(tnow > params.ttl) {
+  if (params.ttl === 0) {
```
Comment (bookchin) [inline lib/quasar.js]: I think we can just change this to `params.ttl < 1` @yuilleb

Comment (yuilleb) [inline lib/quasar.js]: Agreed. That should work

File: lib/quasar.js [hunk 2]
```diff
@@ -216,35 +217,29 @@ Quasar.prototype._sendPublish = function(contact, params, callback) {
...
+    kad.constants.K,
+    this._router._self.nodeID
+  );
 
   // Check to make sure that we have not already seen this message
   if (this._seen.get(params.uuid)) {
     return callback(new Error('Message previously routed'));
   }
 
+  if (params.ttl > this._options.maxRelayHops) {
```
Comment (yuilleb) [inline lib/quasar.js]: @bookchin Could also add a check here making sure it is greater than 0 as well.

Comment (yuilleb) [inline lib/quasar.js]: something like `if (params.ttl > this._options.maxRelayHops || params.ttl < 0)`

Comment (yuilleb) [inline lib/quasar.js]: I suppose there's no harm in processing an invalid ttl message, it's just the forwarding of it we don't want, so up to you

Comment (bookchin) [inline lib/quasar.js]: Yeah might as well check there too. We'd still want to handle a 0 TTL message - that just means we are last in line. Here we can make sure it is `> -1` which should never happen and if it does, we shouldn't handle the message.

**Comment by bookchin** [comment]
https://github.com/kadtools/kad-quasar/releases/tag/v1.0.1

**Pull request merged**